### ISSUE TITLE
Support for Python 3

### ISSUE THREAD

**Issue by trehn** [Yubico/python-u2flib-server#9]
This is currently blocked by the dependency on M2Crypto. Reading the discussion in martinpaljak/M2Crypto#23 is not very encouraging. M2Crypto seems rather poorly maintained in general.

Poking around the source a bit, my impression is that M2Crypto could be replaced with pyopenssl. Can someone more familiar with the project confirm or deny?

**Comment by dainnilsson** [comment]
I took a quick glance at pyopenssl, but it doesn't seem to support any ECC primitives (please correct me if I'm wrong). U2F requires support for the secp256r1 curve at least.

**Comment by trehn** [comment]
Have you also looked at "cryptography" by any chance? https://cryptography.io/en/latest/hazmat/primitives/asymmetric/ec/#cryptography.hazmat.primitives.asymmetric.ec.SECP256R1

**Comment by dainnilsson** [comment]
Very briefly. It looks like a cleaner implementation than M2Crypto, and does seem to support the required curve. The x509 support seems a bit immature though, unfortunately. It isn't even available in versions < 0.7 (which is what most package managers are distributing). The support in the latest version doesn't seem to expose much either, I'm afraid. As a baseline, we'd need to be able to parse two certs, A and B, and be able to verify that A is the signer of B. This needs to work for RSA certificates as well as EC ones. From my quick testing I wasn't able to figure out how to do so, using the 0.8 version of the library.

In summary, it has potential, but isn't quite there yet!

**Comment by gavinwahl** [comment]
This is being worked on in https://github.com/pyca/cryptography/issues/1957

**Comment by moreati** [comment]
I've done a first pass in https://github.com/Yubico/python-u2flib-server/compare/Yubico:2e9fca4...moreati:926d521. Note that

 - I am not a cryptographer, It needs review
 - It's incomplete - at the moment M2Crypto is still necessary to verify issuer certificates

**Comment by dainnilsson** [comment]
Great job so far! I've made a few comments in your changes. Hopefully the last bit of M2Crypto can be replaced soon.

**Comment by dainnilsson** [comment]
pyca/Cryptography#2381 seems like it would take care of the certificate verification problem. I think it can actually be done now already, using pyasn1 to get the signature from a certificate, as well as the part that the hash is calculated from, and then verifying it using ca.public_key().verifier(signature, cert.signature_hash_algorithm).update(cert_data).verify()

The above is untested, and I'm not sure it makes sense to implement this rather than waiting for the proper Cryptography APIs.

**Comment by moreati** [comment]
@dainnilsson A quick sanity check, the current implementation of `MetadataResolver.resolve()` with M2Crypto uses `cert.verify(issuer.get_pubkey())`. That calls OpenSSL's `X509_verify()` under the hood, which AFAICT *only* verifies the signature matches. It ignores e.g. trust chains, expiration time.

1. Is the above accurate?
2. Is the lack of further verification by design (e.g. specified by the U2F spec), or a compromise/MVP?

**Comment by dainnilsson** [comment]
That is accurate. Trust chains are not applicable in this case, as you are explicitly passing the issuer cert to verify the signature against (which is already marked as trusted). Expiration time is in this case ignored, much due to the fact that it is under-specified how it should be handled, as attestation certificates cannot be updated.

The U2F specification has nothing to say on the subject of attestation verification. The implementation in this library is one lightweight approach, which assumes you have some method of vetting which CAs to trust, and for how long.

**Comment by moreati** [comment]
> I think it can actually be done now already, using pyasn1

I can confirm this, [but it's ugly as hell](https://github.com/moreati/python-u2flib-server/commit/8ddf35d0112b257b9b6c50fd564f538cddc50823). I don't plan on creating a PR of that sub-branch, unless you want to. Much cleaner to wait for cryptography to release an implementation.

**Comment by vcarel** [comment]
@moreati I tried your patch for removing M2Crypto. I don't think it is that ugly. I far more annoyed by M2Crypto as of today.

Why not going for certificate validation with pyasn1 now, and then remove the work-around when it's ready in Cryptography ?

**Comment by mark-adams** [comment]
FYI: in PR #17 I've completed all the needed work to swap out M2Crypto for `cryptography` with 100% of tests passing.
